libd/src/d/semantic/symbol.d

```diff
@@ -754,19 +754,79 @@ struct SymbolAnalyzer {
 			methodIndex = oldMethodIndex;
 		}
 
-		thisType = Type.get(i).getParamType(false, true);
-
+		thisType = Type.get(pass.object.getObject()).getParamType(false, true);
 		import std.conv;
 		auto name = i.name.toString(context);
 		manglePrefix = manglePrefix ~ to!string(name.length);
 
 		i.mangle = "I" ~ manglePrefix;
 
-		assert(d.members.length == 0, "Member support not implemented for interfaces yet");
-		assert(d.bases.length == 0, "Interface inheritance not implemented yet");
+		foreach(b; d.bases) {
+			import d.semantic.identifier : AliasResolver;
+			i.bases ~= AliasResolver!(function Interface (identified) {
+					static if(is(typeof(identified) : Symbol)) {
+						if(auto i = cast(Interface) identified) {
+							return i;
+						}
+					}
+					
+					static if(is(typeof(identified.location))) {
+						import d.exception;
+						throw new CompileException(identified.location, typeid(identified).toString() ~ " is not an interface.");
+					} else {
+						// for typeof(null)
+						assert(0);
+					}
+				})(pass).visit(b);
+		}
+
+
+		auto members = DeclarationVisitor(pass, AggregateType.Class).flatten(d.members, i);
+		Method[] methods;
+
+		foreach(m; members) {
+			if (auto method = cast (Method) m) { 
+				scheduler.require(method, Step.Signed);
+				// check for (non-static) method body
+				if(method.fbody) {
+					import d.exception;
+					throw new CompileException(method.location, "non-static or non-final method can't have a body in interface");
+				}
+				methods ~= method;
+
+			} else if(auto staticMethod = cast(Function) m) { // static method
+				// TODO: handle static
+			} else { // not a method
+				import d.exception;
+				throw new CompileException(m.location, "Interface can have only methods");
+			}
+
+		}
+
 
-		// TODO: lots of stuff to add
+		// generate the vtable field 
+		auto vtblType = Type.get(BuiltinType.Void).getPointer(TypeQualifier.Immutable);
+		import d.context.name : BuiltinName;
+		auto vtbl = new Field(d.location, 0, vtblType, BuiltinName!"__vtbl", null);
+		vtbl.step = Step.Processed;
+		i.members ~= [vtbl];
 
+		// add base interface methods
+		scheduler.require(i.bases);
+		foreach(b; i.bases) {
+			foreach(m; b.members) {
+			// TODO: handle static/final methods
+				if (auto vtable = cast(Field) m) {
+					i.members ~= vtable;
+				}else if (auto method = cast(Method) m) {
```

Comment:
Please follow the style. it is not like it is diffucult, and that makes your code easier to read.